service/discovery/aws/storage.go

```
@@ -0,0 +1,382 @@
+		log.Println("Getting resources for", bucket.name)
+		encryptionAtRest, err = d.getEncryptionAtRest(bucket.name)
+		if err != nil {
+			return
+		}
+		encryptionAtTransmit, err = d.getTransportEncryption(bucket.name)
+		if err != nil {
+			return
+		}
+		resources = append(resources, &voc.ObjectStorageResource{
+			StorageResource: voc.StorageResource{
+				Resource: voc.Resource{
+					ID:           bucket.arn,
+					Name:         bucket.name,
+					CreationTime: bucket.creationTime.Unix(),
+					Type:         []string{"ObjectStorage", "Storage", "Resource"},
+				},
+				AtRestEncryption: &encryptionAtRest,
+			},
+			HttpEndpoint: &voc.HttpEndpoint{
+				URL:                 bucket.endpoint,
+				TransportEncryption: &encryptionAtTransmit,
+			},
+		})
+	}
+	return
+}
+func (b bucket) String() string {
+	return fmt.Sprintf("[ARN: %v, Name: %v, Creation Time: %v, Number of objects: %v]", b.arn, b.name, b.creationTime, b.numberOfObjects)
+}
+
+// NewAwsStorageDiscovery constructs a new awsS3Discovery initializing the s3-client and isDiscovering with true
+func NewAwsStorageDiscovery(cfg aws.Config) discovery.Discoverer {
+	return &awsS3Discovery{
+		client:        s3.NewFromConfig(cfg),
+		buckets:       nil,
+		isDiscovering: true,
+	}
+}
+
+// S3ListBucketsAPI is the interface for the List function (used for mock testing)
+type S3ListBucketsAPI interface {
+	ListBuckets(ctx context.Context,
+		params *s3.ListBucketsInput,
+		optFns ...func(*s3.Options)) (*s3.ListBucketsOutput, error)
+}
+
+// getBuckets returns all buckets
+func (d *awsS3Discovery) getBuckets() (err error) {
+	log.Println("Getting buckets in s3...")
+	var resp *s3.ListBucketsOutput
+	resp, err = d.client.ListBuckets(context.TODO(), &s3.ListBucketsInput{})
+	if err != nil {
+		log.Error("Error occurred.")
+		var oe *smithy.OperationError
+		if errors.As(err, &oe) {
+			log.Printf("failed to call service: %s, operation: %s, error: %v", oe.Service(), oe.Operation(), oe.Unwrap())
```

rather than log printing it here, you could return a new error with `fmt.Errorf` that contains additional details where the error occurred.

---

Right, good hint. that would also help to keep aws stuff out of the discoverer. I will look into that.